### ISSUE TITLE
Disconnect / Re-connect

### ISSUE THREAD

**Issue by NS-K** [TheRealMazur/LinakDeskEsp32Controller#12]
with linak desk you can only have one Bluetooth device connected to the desk. to go around this i have added a timer that make the esp32 disconnect from the desk by calling `controller.disconnect()`. but i am not able to make it connect back until rebooting the esp32. do you have any idea what can cause this ?

**Comment by TheRealMazur** [comment]
Hey, I have tested disconnecting and reconnecting in a loop on my local setup and it works fine for me. However there might be problems with the bonding information on the desk controller side when you pair your phone. Do you have a fork of the repo that I could look at to better reproduce it?

**Comment by NS-K** [comment]
My linak have a ble remote control and that's the reason why i want to disconnect the esp32 from the linak controller.
so during the test i ve removed the battery from the remote control to be sure that it's not affecting the result.
I have also tried to compile the code with `h2zero/NimBLE-Arduino @ 1.3.6` but the same problem remain.
I have also tried to deinit the ble after disconnecting
I will give a try tomorrow to the `h2zero/NimBLE-Arduino @ 1.3.7` as they fixed some Memory leak when deleting client instances.
I do not have a fork of the repo but this is the change that i made to your main.cpp.

```
#define DISCONNECT_TIMEOUT 30000

bool linakConnect() {
	
	if (!controller.isConnected()) {
		controller.connect(BtAddress);
	}

	if (controller.isConnected()) {		
		connectTime = millis();
	}
	
	return controller.isConnected();

}

//  on DeskController.cpp i added
// bool DeskController::isMoving() { return mIsMoving; }

bool linakDisconnect() {

	if(!controller.isMoving()){
		controller.disconnect();		
	}
		
	return !controller.isConnected();
}

// on the loop()

if(millis() - connectTime > DISCONNECT_TIMEOUT && controller.isConnected()) {

		linakDisconnect();
}
```

**Comment by TheRealMazur** [comment]
Okay, first of all, if you don't want to create a fork of the repo, please use the output of `git diff` to share ***all*** the changes you made, otherwise I cannot know if something else is not affecting the code.  
I.e. in the code you've shared I do not see a call to `linakConnect` so it seems normal that the controller doesn't reconnect after the first disconnect 😜    
Also, please try describing what you are trying to accomplish a bit more. I assumed you want to use another device (like a smartphone) to control the desk while the ESP32 is disconnected. However after your comment about taking the battery out of some controller, I'm a bit confused about the setup you have.

**Comment by NS-K** [comment]
i have a linak actuator that is controlled only by a remote control and what i am working on is to create a esp32-homekit that will connect to this actuator and will be integrated into one of my homekit scene. and this is the reason why i didn't share my complete code with you as i removed the DoubleResetDetector, AsyncWebServer, fauxmoESP ... and added the esp32-homekit library.
everything is working fine except the remote control stop working which is normal because only one ble device could be connected to the controller.
so i come with the idea to add a timeout to disconnect the esp32-homekit from the linak and only connected it back when a command is sent from homekit `if (linakConnect() && destination >= 0 && destination < MAX_HEIGHT_RAW) { controller.moveToHeightRaw(destination); }` but it always end with a `"NimBLEClient:" Connection failed; status=534`.

**Comment by TheRealMazur** [comment]
Oh, I see. I think the problem lies in the instability of having both WiFi and BLE running at the same time. I'm guessing (again, because of the lack of the actual code) that you are doing `linakConnect()` during handling of an HTTP call, so before the HTTP response is sent.  I was seeing some of those problems myself when developing the project. That's why I made all the `moveToHeight` commands non blocking. ESP32 modules use the same radio for both BLE and WiFi so I don't think establishing an BLE connection while handling an HTTP request is achievable.  

From the maintainer of NimBLE-Arduino lib I got the [info](https://github.com/h2zero/NimBLE-Arduino/issues/323#issuecomment-1002236218) that:  
> BLE+async are a known problematic combination

I'm actually starting to work on another project due to that limitation - I want to use my Core library and emulate a BLE smarthome device for the smart speakers to connect directly - so that the WiFi stack doesn't need to be used on the ESP32. That will bring it's own limitations, as all configuration will most likely have to be done in compile time. I'm not sure how is the HomeKit BLE support, but that might be one way forward for you if the WiFi + Bluetooth combo won't work 🙂
